internal/custom_types/qosDppPol_violateMarkDscp.go

```diff
@@ -0,0 +1,149 @@
+
+type QosDppPolViolateMarkDscpStringType struct {
+	basetypes.StringType
+}
+
+func (t QosDppPolViolateMarkDscpStringType) Equal(o attr.Type) bool {
+	other, ok := o.(QosDppPolViolateMarkDscpStringType)
+
+	if !ok {
+		return false
+	}
+
+	return t.StringType.Equal(other.StringType)
+}
+
+func (t QosDppPolViolateMarkDscpStringType) String() string {
+	return "QosDppPolViolateMarkDscpStringType"
+}
+
+func (t QosDppPolViolateMarkDscpStringType) ValueFromString(ctx context.Context, in basetypes.StringValue) (basetypes.StringValuable, diag.Diagnostics) {
+	value := QosDppPolViolateMarkDscpStringValue{
+		StringValue: in,
+	}
+
+	return value, nil
+}
+
+func (t QosDppPolViolateMarkDscpStringType) ValueFromTerraform(ctx context.Context, in tftypes.Value) (attr.Value, error) {
+	attrValue, err := t.StringType.ValueFromTerraform(ctx, in)
+
+	if err != nil {
+		return nil, err
+	}
+
+	stringValue, ok := attrValue.(basetypes.StringValue)
+
+	if !ok {
+		return nil, fmt.Errorf("unexpected value type of %T", attrValue)
+	}
+
+	stringValuable, diags := t.ValueFromString(ctx, stringValue)
+
+	if diags.HasError() {
+		return nil, fmt.Errorf("unexpected error converting StringValue to StringValuable: %v", diags)
+	}
+
+	return stringValuable, nil
+}
+
+func (t QosDppPolViolateMarkDscpStringType) ValueType(ctx context.Context) attr.Value {
+	return QosDppPolViolateMarkDscpStringValue{}
+}
+
+// QosDppPolViolateMarkDscp custom string value.
+
+var _ basetypes.StringValuableWithSemanticEquals = QosDppPolViolateMarkDscpStringValue{}
+
+type QosDppPolViolateMarkDscpStringValue struct {
+	basetypes.StringValue
+}
+
+func (v QosDppPolViolateMarkDscpStringValue) Equal(o attr.Value) bool {
+	other, ok := o.(QosDppPolViolateMarkDscpStringValue)
+
+	if !ok {
+		return false
+	}
+
+	return v.StringValue.Equal(other.StringValue)
+}
+
+func (v QosDppPolViolateMarkDscpStringValue) Type(ctx context.Context) attr.Type {
+	return QosDppPolViolateMarkDscpStringType{}
+}
+
+func (v QosDppPolViolateMarkDscpStringValue) StringSemanticEquals(ctx context.Context, newValuable basetypes.StringValuable) (bool, diag.Diagnostics) {
+	var diags diag.Diagnostics
+
+	newValue, ok := newValuable.(QosDppPolViolateMarkDscpStringValue)
+
+	if !ok {
+		diags.AddError(
+			"Semantic Equality Check Error",
+			"An unexpected value type was received while performing semantic equality checks. "+
+				"Please report this to the provider developers.\n\n"+
+				"Expected Value Type: "+fmt.Sprintf("%T", v)+"\n"+
+				"Got Value Type: "+fmt.Sprintf("%T", newValuable),
+		)
+
+		return false, diags
+	}
+
+	priorMappedValue := QosDppPolViolateMarkDscpValueMap(v.StringValue)
+
+	newMappedValue := QosDppPolViolateMarkDscpValueMap(newValue.StringValue)
+
+	return priorMappedValue.Equal(newMappedValue), diags
+}
+
+func QosDppPolViolateMarkDscpValueMap(value basetypes.StringValue) basetypes.StringValue {
+	matchMap := map[string]string{
+		"0xffff": "unspecified",
```

Comment:
this is actually necessary as this semantic equality is even mentioned in the UI.
<img width="1512" alt="Screenshot 2024-08-21 at 11 26 36 AM" src="https://github.com/user-attachments/assets/ccc8704b-17b1-4dc3-91b0-654e3b73072d">

Reply:
This might actually be useful in the future. 
We discovered that we can actually use hexadecimal values as input for a few attributes. "flag" attributes use it as-well for RGB colors for example, but it doesn't satisfy the current conditions to receive a customtype (they don't have a validator list in their meta). 
It's too soon to know what might be the exact conditions for adding a custom type to an attribute. I think the best thing for now is to wait till we generate more resources (and start a refactor of the generator). 
This PR only focuses on semantic equality between float and string (for example TCP port values which was the original issue). 
We might need to extend the customtype generation and conditions in the future. But for now, I think it is best to stick to this solution.

Reply:
Sounds good to me